nemo/collections/nlp/models/language_modeling/megatron/gpt_full_te_layer_autocast_spec.py

```diff
@@ -15,14 +15,26 @@
 from typing import Any, Callable, Optional
 
 import torch
-from megatron.core import parallel_state, tensor_parallel
-from megatron.core.transformer.spec_utils import ModuleSpec
-from megatron.core.transformer.utils import make_sharded_tensors_for_checkpoint
-
-from transformer_engine.pytorch import TransformerLayer
 
+from nemo.collections.nlp.modules.common.megatron.utils import ApexGuardDefaults
 from nemo.collections.nlp.parts import utils_funcs
 
+try:
+    from megatron.core import parallel_state, tensor_parallel
+    from megatron.core.transformer.spec_utils import ModuleSpec
+    from megatron.core.transformer.transformer_layer import BaseTransformerLayer
+    from megatron.core.transformer.utils import make_sharded_tensors_for_checkpoint
+
+    HAVE_MEGATRON_CORE = True
+
+except (ImportError, ModuleNotFoundError):
+
+    BaseTransformerLayer = ApexGuardDefaults
+
+    HAVE_MEGATRON_CORE = False
+
```

Comment:
@ericharper what do you think?

Reply:
That would do it. But I like better the idea of just capturing the error and then raising an error while logging the captured error when instantiating the class.  If we check based on versioning then, we might still run into issues because, for example, the megatron version hasn't been updated yet

Reply:
It's not just logging, as long as we have a test that instantiates the class, an error will be raised

Reply:
But a different error will be raised during init of TETransformerLayerAutocast making it hard to debug the root cause. So if there is an issue with importing, then that's what needs to show up and then fail right? Not something else later on.

Reply:
If I understand correctly: the problem is that this code will fail in CI, because this feature was added in the recent mcore commit that is not tested in nemo yet. Updating mcore commit  is not always an option because it can break other unrelated things. @ericharper is my understanding correct?

If so, then I think that it is actually ok that CI fails, because it shows that we have an issue somewhere. I think that good solution to this problem is not to add snippets that obscure error message, but rather to improve development process by introducing devel branches that support latest features. And once a dependency is ready (new dependency release) we could merge it to the main branch, marking it stable.